earth2studio/data/goes.py

```diff
@@ -0,0 +1,659 @@
+            with open(cache_path, "wb") as file:
+                await asyncio.to_thread(file.write, data)
+
+        return cache_path
+
+    @property
+    def cache(self) -> str:
+        """Return appropriate cache location."""
+        cache_location = os.path.join(datasource_cache_root(), "goes")
+        if not self._cache:
+            cache_location = os.path.join(cache_location, "tmp_goes")
+        return cache_location
+
+    @classmethod
+    def available(
+        cls,
+        time: datetime | np.datetime64,
+        satellite: str = "goes16",
+        scan_mode: str = "F",
+    ) -> bool:
+        """Checks if given date time is available in the GOES object store
+
+        Parameters
+        ----------
+        time : datetime | np.datetime64
+            Date time to access
+        satellite : str, optional
+            Which GOES satellite to check, by default "goes16"
+        scan_mode : str, optional
+            Which scan mode to check, by default "F"
+
+        Returns
+        -------
+        bool
+            If date time is available
+        """
+        if isinstance(time, np.datetime64):  # np.datetime64 -> datetime
+            _unix = np.datetime64(0, "s")
+            _ds = np.timedelta64(1, "s")
+            time = datetime.fromtimestamp((time - _unix) / _ds, timezone.utc)
+
+        # Check if data exists in S3
+        fs = s3fs.S3FileSystem(anon=True)
+
+        # Get needed date components
+        year = time.year
+        day_of_year = time.timetuple().tm_yday
+        hour = time.hour
+
+        # Construct the base URL
+        base_url = cls.BASE_URL.format(
+            satellite=satellite,
+            scan_mode=scan_mode[0:1],
+            year=year,
+            day_of_year=day_of_year,
+            hour=hour,
+        )
+
+        try:
+            # List files in the directory
+            files = fs.ls(base_url)
+
+            # Filter for files matching the product and scan mode
+            pattern = f"OR_ABI-L2-MCMIP{scan_mode}"
+            matching_files = [f for f in files if pattern in f]
+
+            if not matching_files:
+                return False
+
+            # Sort files by time (same logic as _get_s3_path)
+            def get_time(file_name: str) -> datetime:
+                start_str = file_name.split("/")[-1].split("_")[-3][1:-1]
+                t = datetime.strptime(start_str, "%Y%j%H%M%S")
+                if time.tzinfo is not None:
+                    t = t.replace(tzinfo=timezone.utc)
+                return t
+
+            time_stamps = [get_time(f) for f in matching_files]
+
+            # Get the index of the file that is the closest to the requested time
+            file_index = np.argmin(np.abs(np.array(time_stamps) - time))
+
+            # Check if the specific file exists
+            try:
+                matching_files[file_index]
+                return True
+            except IndexError:
+                return False
+
+        except Exception:
```

Comment:
Are there certain exceptions that we would expect to throw when not available here?
Or Would the one IndexError catch.

Basically if something else is failing (like user has invalid satellite, it would just return fail rather than an error telling them they have a bad input)

Reply:
Took this out and refactored the exceptions to make sure that the satellite and scan mode are ok.